cosmoscx/src/partition_key.rs

```diff
@@ -0,0 +1,113 @@
+};
+use serde_json::Value;
+
+use crate::{
+    result::{FfiResult, ResultExt},
+    slice::{OwnedString, Str},
+};
+
+/// Parses a JSON value into a vector of `PartitionKeyValue`s accepted by the hashing function.
+fn parse_partition_key_components(
+    v: Value,
+) -> Result<Vec<PartitionKeyValue>, azure_data_cosmos_engine::Error> {
+    fn convert(val: &Value) -> Result<PartitionKeyValue, azure_data_cosmos_engine::Error> {
+        Ok(match val {
+            Value::Null => PartitionKeyValue::Null,
+            Value::Bool(b) => PartitionKeyValue::Bool(*b),
+            Value::Number(n) => {
+                PartitionKeyValue::Number(n.as_f64().ok_or_else(|| {
+                    ErrorKind::DeserializationError.with_message("invalid number")
+                })?)
+            }
+            Value::String(s) => PartitionKeyValue::String(s.clone()),
+            Value::Object(map) => {
+                // Only the empty JSON object is allowed and means Undefined / Missing PK component
+                if map.is_empty() {
+                    PartitionKeyValue::Undefined
+                } else {
+                    return Err(ErrorKind::DeserializationError
+                        .with_message("non-empty object not allowed in partition key"));
+                }
+            }
+            Value::Array(_) => {
+                return Err(ErrorKind::DeserializationError
+                    .with_message("nested arrays not allowed in partition key"));
+            }
+        })
+    }
+
+    match v {
+        Value::String(ref s) if s == "Infinity" => {
+            // Special case: top-level "Infinity" represents the boundary value
+            Ok(vec![PartitionKeyValue::Infinity])
+        }
+        Value::Array(arr) => {
+            let mut out = Vec::with_capacity(arr.len());
+            for item in arr.iter() {
+                out.push(convert(item)?);
+            }
+            Ok(out)
+        }
+        // Allow passing a single value instead of an array; wrap it.
+        other => Ok(vec![convert(&other)?]),
+    }
+}
+
+fn inner_compute<'a>(
+    json: Str<'a>,
+    version: u8,
+    kind: u8,
+) -> Result<Box<OwnedString>, azure_data_cosmos_engine::Error> {
+    let json_str = unsafe { json.as_str().not_null()? };
+    // Empty string not allowed (would imply none)
+    if json_str.is_empty() {
+        return Err(ErrorKind::DeserializationError.with_message("empty partition key json"));
+    }
+    let value: Value = serde_json::from_str(json_str)
+        .map_err(|e| ErrorKind::DeserializationError.with_source(e))?;
+    let components = parse_partition_key_components(value)?;
+
+    let pk_kind = match kind {
+        0 => PartitionKeyKind::Hash,
+        1 => PartitionKeyKind::MultiHash,
+        _ => {
+            return Err(ErrorKind::DeserializationError
+                .with_message("invalid partition key kind (expected 0 Hash, 1 MultiHash)"))
+        }
```

Comment:
I like this pattern, since we properly handle invalid values, but we should define public constants for these in this file. If we do that, cbindgen should propagate them to the C header, and callers that use that header file will be able to use the constants. I would also update the message so it doesn't report the possible values, since that should be something we put in the documentation.